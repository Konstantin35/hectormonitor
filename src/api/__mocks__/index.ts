const poolApi = {
  coins: [],
  isCoinSupported: jest.fn(() => true),
  isAdressValid: jest.fn(() => Promise.resolve(true)),
};

export const getPoolApi = jest.fn(() => poolApi);
