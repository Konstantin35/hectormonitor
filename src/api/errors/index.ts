// tslint:disable max-classes-per-file

export class InvalidAddressError extends Error {
  constructor(message: string = 'Invalid address') {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class NoDataAddressError extends Error {
  constructor(message: string = 'Address returns no data for the pool') {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class ApiError extends Error {
  constructor(message: string = 'Api returned an error') {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
