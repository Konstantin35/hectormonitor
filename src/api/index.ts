export { getPoolApi } from './get-pool-api';
