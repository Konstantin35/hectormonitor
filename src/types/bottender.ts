/**
 * bottender context
 */
export interface Context {
  sendText: (message: string, options?: any) => Promise<void>;
  event: {
    isText: boolean;
    text: string;
    message: {
      message_id: string;
      text: string;
      from: {
        id: string;
        is_bot: boolean;
        first_name: string;
        username: string;
        language_code: string;
      };
    };
  };
}
