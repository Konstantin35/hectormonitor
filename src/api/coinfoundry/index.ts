import { sum } from 'lodash';
import { Coins } from '../../types/types';
import { logger } from '../../logger';
import { ApiError } from '../errors';
import { PoolApi, fetch, PoolStats } from '../utils';

interface CoinfoundryStatsWorkers {
  hashrate: number;
  sharesPerSecond: number;
}

interface CoinfoundryStats {
  pendingBalance: number;
  performance: {
    created: string;
    workers: {
      [name: string]: CoinfoundryStatsWorkers;
    };
  };
}

export class Coinfoundry implements PoolApi {
  public apiUrl = 'https://coinfoundry.org/api';
  public coins: Coins[] = [
    Coins.eth,
    Coins.etc,
    Coins.xmr,
    Coins.ltc,
    Coins.dash,
  ];

  public isCoinSupported(coin: Coins) {
    return this.coins.includes(coin);
  }

  public async isAdressValid(coin: Coins, address: string): Promise<boolean> {
    const stats: CoinfoundryStats = await this.fetch(
      `/pool/${this.convertCoinKey(coin)}/account/${address}?perfMode=day`
    );
    if (!stats.performance) {
      return false;
    }
    return true;
  }

  public async getStats(coin: Coins, address: string): Promise<PoolStats> {
    const stats: CoinfoundryStats = await this.fetch(
      `/pool/${this.convertCoinKey(coin)}/account/${address}?perfMode=day`
    );
    return this.normalizeStats(stats);
  }

  private normalizeStats(stats: CoinfoundryStats): PoolStats {
    // Need to sum all workers hashrate
    const currentHashrate = sum(
      Object.keys(stats.performance.workers).map(
        (key: string) => stats.performance.workers[key].hashrate
      )
    );

    return {
      currentHashrate,
      unpaid: stats.pendingBalance || 0,
      lastSeen: new Date(`${stats.performance.created}z`).getTime(),
      activeWorkers: Object.keys(stats.performance.workers).length,
    };
  }

  private async fetch(endpoint: string): Promise<any> {
    const res = await fetch(`${this.apiUrl}${endpoint}`);
    const { result, success } = res;
    if (!success) {
      logger.error(res);
      throw new ApiError();
    }
    return result;
  }

  private convertCoinKey(coin: Coins): string {
    return `${coin}1`;
  }
}
