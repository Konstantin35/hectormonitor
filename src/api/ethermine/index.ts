import * as ethunits from 'ethereum-units';
import { InvalidAddressError, NoDataAddressError, ApiError } from '../errors';
import { Coins } from '../../types/types';
import { logger } from '../../logger';
import { PoolApi, PoolPayout, fetch, PoolStats } from '../utils';

export class Ethermine implements PoolApi {
  public apiUrl = 'https://api.ethermine.org';
  public coins: Coins[] = [Coins.eth];

  public isCoinSupported(coin: Coins) {
    return this.coins.includes(coin);
  }

  public async isAdressValid(_: Coins, address: string): Promise<boolean> {
    try {
      await this.fetch(`/miner/${address}/currentStats`);
    } catch (err) {
      if (err instanceof NoDataAddressError) {
        return false;
      }
      throw err;
    }
    return true;
  }

  public async getLastPayout(
    _: Coins,
    address: string
  ): Promise<PoolPayout | null> {
    const payouts: any[] = await this.fetch(`/miner/${address}/payouts`);
    return payouts && payouts.length > 0
      ? this.normalizePayout(payouts[0])
      : null;
  }

  public async getPayouts(_: Coins, address: string): Promise<PoolPayout[]> {
    const payouts: any[] = await this.fetch(`/miner/${address}/payouts`);
    return payouts.map(this.normalizePayout);
  }

  public async getStats(_: Coins, address: string): Promise<PoolStats> {
    const stats = await this.fetch(`/miner/${address}/currentStats`);
    return this.normalizeStats(stats);
  }

  private normalizePayout(payout: PoolPayout): PoolPayout {
    return {
      txHash: payout.txHash,
      amount: Number(ethunits.convert(payout.amount, 'wei', 'ether')),
      paidOn: payout.paidOn * 1000,
    };
  }

  private normalizeStats(stats: PoolStats): PoolStats {
    return {
      reportedHashrate: stats.reportedHashrate,
      currentHashrate: stats.currentHashrate || 0,
      averageHashrate: stats.averageHashrate,
      unpaid: Number(ethunits.convert(stats.unpaid || 0, 'wei', 'ether')),
      lastSeen: stats.lastSeen * 1000,
      activeWorkers: stats.activeWorkers || 0,
    };
  }

  private async fetch(endpoint: string): Promise<any> {
    const res = await fetch(`${this.apiUrl}${endpoint}`);
    const { data, status, error } = res;
    // Api return error
    if (status === 'ERROR') {
      if (error === 'Invalid address') {
        throw new InvalidAddressError();
      }
      logger.error(res);
      throw new ApiError();
    }
    // Api return no data
    if (data === 'NO DATA') {
      throw new NoDataAddressError();
    }

    // No data probably api not working well
    if (!data) {
      logger.error(res);
      throw new ApiError();
    }
    return data;
  }
}
