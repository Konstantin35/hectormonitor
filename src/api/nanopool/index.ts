import { Coins } from '../../types/types';
import { PoolApi, fetch, PoolStats, PoolPayout } from '../utils';
import { ApiError } from '../errors';
import { logger } from '../../logger';

interface NanopoolPayout {
  date: number;
  txHash: string;
  amount: number;
  confirmed: boolean;
}

interface NanopoolUser {
  balance: number;
  hashrate: number;
  workers: any[];
  avgHashrate: {
    h24: number;
  };
}

export class Nanopool implements PoolApi {
  public apiUrl = 'https://api.nanopool.org/v1';
  public coins: Coins[] = [Coins.eth, Coins.etc, Coins.xmr];

  public isCoinSupported(coin: Coins) {
    return this.coins.includes(coin);
  }

  public async isAdressValid(coin: Coins, address: string): Promise<boolean> {
    try {
      const res = await this.fetch(coin, `/accountexist/${address}`);
      if (res === 'Account exist') {
        return true;
      }
    } catch (err) {
      if (err === 'Account not found') {
        return false;
      }
    }
    return false;
  }

  public async getStats(coin: Coins, address: string): Promise<PoolStats> {
    const user = await this.fetch(coin, `/user/${address}`);
    return this.normalizeStats(user);
  }

  public async getLastPayout(
    coin: Coins,
    address: string
  ): Promise<PoolPayout | null> {
    const payouts: any[] = await this.fetch(coin, `/payments/${address}`);
    return payouts && payouts.length > 0
      ? this.normalizePayout(payouts[0])
      : null;
  }

  public async getPayouts(coin: Coins, address: string): Promise<PoolPayout[]> {
    const payouts: any[] = await this.fetch(coin, `/payments/${address}`);
    return payouts.map(this.normalizePayout);
  }

  private normalizePayout(payout: NanopoolPayout): PoolPayout {
    return {
      txHash: payout.txHash,
      amount: payout.amount,
      paidOn: payout.date * 1000,
    };
  }

  private normalizeStats(stats: NanopoolUser): PoolStats {
    return {
      currentHashrate: Number(stats.hashrate) || 0,
      averageHashrate: Number(stats.avgHashrate.h24) || 0,
      unpaid: Number(stats.balance),
      activeWorkers: stats.workers.length || 0,
    };
  }

  private async fetch(coin: Coins, endpoint: string): Promise<any> {
    const res = await fetch(`${this.apiUrl}/${coin}${endpoint}`);
    const { data, error } = res;
    if (error) {
      logger.error(res);
      throw new ApiError(error);
    }
    return data;
  }
}
