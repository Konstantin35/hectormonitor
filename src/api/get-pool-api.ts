import { Pools } from '../types/types';
import { PoolApi } from './utils';
import { Ethermine } from './ethermine';
import { Nanopool } from './nanopool';
import { Coinfoundry } from './coinfoundry';

const map = {
  [Pools.ethermine]: new Ethermine(),
  [Pools.nanopool]: new Nanopool(),
  [Pools.coinfoundry]: new Coinfoundry(),
};

export const getPoolApi = (pool: Pools): PoolApi => {
  return map[pool];
};
