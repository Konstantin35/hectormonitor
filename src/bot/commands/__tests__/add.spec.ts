import { ContextSimulator } from 'bottender/test-utils';
import mockingoose from 'mockingoose';
import { handler } from '../../index';
import { getPoolApi } from '../../../api';

jest.mock('../../../api');

const mockedPoolApi = getPoolApi(null as any);

const simulator = new ContextSimulator({
  platform: 'telegram',
});
const from = {
  id: 42,
};

describe('add command', () => {
  beforeEach(() => {
    mockingoose.resetAll();
    jest.clearAllMocks();
  });

  it('should reply invalid message when no arguments', async () => {
    const context = simulator.createTextContext('/add');
    await handler(context);
    expect(context.sendText).toMatchSnapshot();
  });

  it('should reply invalid message when no enough arguments', async () => {
    const context = simulator.createTextContext('/add eth');
    await handler(context);
    expect(context.sendText).toMatchSnapshot();
  });

  it('should reply invalid message when invalid coin', async () => {
    const context = simulator.createTextContext('/add bla bla bla');
    await handler(context);
    expect(context.sendText).toMatchSnapshot();
  });

  it('should reply invalid message when invalid pool', async () => {
    const context = simulator.createTextContext('/add eth bla bla');
    await handler(context);
    expect(context.sendText).toMatchSnapshot();
  });

  it('should reply invalid message when coin is not supported', async () => {
    (mockedPoolApi.isCoinSupported as jest.Mock).mockReturnValueOnce(false);
    const context = simulator.createTextContext('/add etc ethermine bla');
    context.from = from;
    await handler(context);
    expect(getPoolApi).toBeCalledWith('ethermine');
    expect(mockedPoolApi.isCoinSupported).toBeCalledWith('etc');
    expect(context.sendText).toMatchSnapshot();
  });

  it('should reply invalid message when invalid address from api', async () => {
    (mockedPoolApi.isAdressValid as jest.Mock).mockReturnValueOnce(false);
    const context = simulator.createTextContext(
      '/add eth ethermine ivalidAddress'
    );
    context.from = from;
    await handler(context);
    expect(getPoolApi).toBeCalledWith('ethermine');
    expect(mockedPoolApi.isAdressValid).toBeCalledWith('eth', 'ivalidAddress');
    expect(context.sendText).toBeCalledWith('Invalid address');
  });

  it('should create a new user if it does not exist', async () => {
    mockingoose.User.toReturn(null, 'findOne');
    const context = simulator.createTextContext(
      '/add eth ethermine ivalidAddress'
    );
    context.event.message.from = from;
    await handler(context);
    expect(getPoolApi).toBeCalledWith('ethermine');
    expect(mockedPoolApi.isAdressValid).toBeCalledWith('eth', 'ivalidAddress');
    // TODO find a way to see that user.save has been called
  });

  it('should reply invalid message is address already exist', async () => {
    mockingoose.User.toReturn(null, 'findOne');
    mockingoose.Address.toReturn({}, 'findOne');
    const context = simulator.createTextContext(
      '/add eth ethermine ivalidAddress'
    );
    context.event.message.from = from;
    await handler(context);
    expect(mockedPoolApi.isAdressValid).toBeCalledWith('eth', 'ivalidAddress');
    expect(context.sendText).toMatchSnapshot();
  });

  it('should create an address and reply success message', async () => {
    const context = simulator.createTextContext(
      '/add eth ethermine ivalidAddress'
    );
    context.event.message.from = from;
    await handler(context);
    expect(mockedPoolApi.isAdressValid).toBeCalledWith('eth', 'ivalidAddress');
    expect(context.sendText).toMatchSnapshot();
  });
});
