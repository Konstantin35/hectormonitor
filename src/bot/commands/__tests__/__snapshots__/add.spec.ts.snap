// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add command should create an address and reply success message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "
Thanks, you're all set. Enjoy ! üöÄ
      ",
    ],
  ],
}
`;

exports[`add command should reply invalid message is address already exist 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "This address is already registered, please choose another one",
    ],
  ],
}
`;

exports[`add command should reply invalid message when coin is not supported 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "This coin is not suppoorted on the pool",
    ],
  ],
}
`;

exports[`add command should reply invalid message when invalid coin 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Invalid coin (supported coins: eth etc xmr ltc dash)",
    ],
  ],
}
`;

exports[`add command should reply invalid message when invalid pool 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Invalid pool (supported pools: ethermine coinfoundry)",
    ],
  ],
}
`;

exports[`add command should reply invalid message when no arguments 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "
‚ö†Ô∏è *Invalid arguments*
-----------------
/start <coin name> <pool name> <address>
-----------------
Valid coins: eth etc xmr ltc dash
Valid pools: ethermine coinfoundry
_Example: \\"/start eth ethermine 0x00000000...\\"_
    ",
      Object {
        "parse_mode": "Markdown",
      },
    ],
  ],
}
`;

exports[`add command should reply invalid message when no enough arguments 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "
‚ö†Ô∏è *Invalid arguments*
-----------------
/start <coin name> <pool name> <address>
-----------------
Valid coins: eth etc xmr ltc dash
Valid pools: ethermine coinfoundry
_Example: \\"/start eth ethermine 0x00000000...\\"_
    ",
      Object {
        "parse_mode": "Markdown",
      },
    ],
  ],
}
`;
